src/clean_air/models.py

```diff
@@ -27,21 +34,879 @@ class ContactDetails:
+            days: Optional[float] = None,
+            hours: Optional[float] = None,
+            minutes: Optional[float] = None,
+            seconds: Optional[float] = None,
+            *,  # Prevent weeks from being specified by positional argument, because it's invalid to combine weeks
+            # with other arguments
+            weeks: Optional[float] = None,
+    ):
+        # Remember that float implies int as well, so either can be stored, and its type matters for reconstructing
+        # parsed strings that match the original input (with the caveat that we don't store which decimal separator
+        # was used, as that seems too esoteric and fiddly for the uses this code was written for)
+
+        args = [years, months, days, hours, minutes, seconds, weeks]
+        if all(arg is None for arg in args):
+            raise ValueError("At least one argument must be supplied")
+        if any(arg < 0 for arg in args if arg is not None):
+            raise ValueError(f"Arguments cannot be less than 0! arguments={args!r}")
+
+        if weeks is not None and any(v is not None for v in args[:-1]):
+            # If we allowed this, we couldn't generate a valid ISO 8601 string
+            raise ValueError("'weeks' cannot be combined with any other arguments")
+
+        if months and not isinstance(months, int):
+            raise ValueError(
+                "'months' must be an int because non-integer years and months are ambiguous and not currently supported"
+            )
+        self._years = years
+        self._months = months
+        self._days = days
+        self._hours = hours
+        self._minutes = minutes
+        self._seconds = seconds
+        self._weeks = weeks
+
+    @property
+    def years(self) -> float:
+        return self._years if self._years is not None else 0
+
+    @property
+    def months(self) -> int:
+        return self._months if self._months is not None else 0
+
+    @property
+    def days(self) -> float:
+        return self._days if self._days is not None else 0
+
+    @property
+    def hours(self) -> float:
+        return self._hours if self._hours is not None else 0
+
+    @property
+    def minutes(self) -> float:
+        return self._minutes if self._minutes is not None else 0
+
+    @property
+    def seconds(self) -> float:
+        return self._seconds if self._seconds is not None else 0
+
+    @property
+    def weeks(self) -> float:
+        return self._weeks if self._weeks is not None else 0
+
+    @cached_property
+    def _id(self):
+        """
+        This is a helper method that standardises how we compare Duration instances.
+        It's used in functions like __hash__, __eq__, __lte__, etc
+
+        Cached because it requires various calculations and has the potential to be called a lot.
+        (We can cache it because Duration is designed to be immutable)
+        """
+        return Duration._get_months_seconds(**Duration._to_dict(self))
```

Comment:
I like to be explicit when I'm accessing static methods, otherwise, there's not any indication to the reader that it is a static method unless they look up the definition